### PR TITLE
ide : Disallow renaming of non-local items

### LINKED ISSUE #14850: Should not rename structs in standard library
I accidentally renamed `Option::None` to `NoOp`. Although I closed my editor (Neovim) immediately, the release build was failing.

<details>

```sh
$ cargo b --release
   Compiling aho-corasick v1.0.1
   Compiling errno v0.3.1
   Compiling io-lifetimes v1.0.10
   Compiling syn v1.0.109
   Compiling log v0.4.17
   Compiling once_cell v1.17.1
   Compiling termcolor v1.2.0
error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.17.1/src/lib.rs:444:47
    |
444 |             OnceCell { inner: UnsafeCell::new(NoOp) }
    |                                               ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/once_cell-1.17.1/src/lib.rs:936:17
    |
936 |                 NoOp
    |                 ^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `once_cell` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:589:18
    |
589 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:768:18
    |
768 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:943:18
    |
943 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:927:24
    |
927 |             separator: NoOp,
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:958:18
    |
958 |             _ => NoOp,
    |                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1662:23
     |
1662 |             fg_color: NoOp,
     |                       ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1058:30
     |
1058 |                 module_path: NoOp,
     |                              ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1663:23
     |
1663 |             bg_color: NoOp,
     |                       ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1844:25
     |
1844 |         self.fg_color = NoOp;
     |                         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/termcolor-1.2.0/src/lib.rs:1845:25
     |
1845 |         self.bg_color = NoOp;
     |                         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1059:23
     |
1059 |                 file: NoOp,
     |                       ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.17/src/lib.rs:1060:23
     |
1060 |                 line: NoOp,
     |                       ^^^^ not found in this scope

error: could not compile `log` due to 7 previous errors
error: could not compile `termcolor` due to 5 previous errors
error[E0408]: variable `NoOp` is not bound in all patterns
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:251:20
    |
251 |             NoOp | Some(ForceAlgorithm::Teddy) => {
    |             ----   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ pattern doesn't bind `NoOp`
    |             |
    |             variable not in all patterns
    |
help: if you meant to match on a variant or a `const` item, consider making the path in the pattern qualified: `path::to::ModOrType::NoOp`
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:251:13
    |
251 |             NoOp | Some(ForceAlgorithm::Teddy) => {
    |             ^^^^

error[E0416]: identifier `NoOp` is bound more than once in the same pattern
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:207:20
    |
207 |             (NoOp, NoOp) if self.ascii_case_insensitive => NoOp,
    |                    ^^^^ used in a pattern more than once

error[E0416]: identifier `NoOp` is bound more than once in the same pattern
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:208:20
    |
208 |             (NoOp, NoOp) => {
    |                    ^^^^ used in a pattern more than once

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:816:33
    |
816 |         OverlappingState { mat: NoOp, id: NoOp, at: 0, next_match_index: NoOp }
    |                                 ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:816:43
    |
816 |         OverlappingState { mat: NoOp, id: NoOp, at: 0, next_match_index: NoOp }
    |                                           ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:816:74
    |
816 |         OverlappingState { mat: NoOp, id: NoOp, at: 0, next_match_index: NoOp }
    |                                                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:869:51
    |
869 |         Ok(FindIter { aut, input, last_match_end: NoOp })
    |                                                   ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1172:32
     |
1172 |                         return NoOp;
     |                                ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1209:9
     |
1209 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1224:9
     |
1224 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1235:9
     |
1235 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1264:19
     |
1264 |         return Ok(NoOp);
     |                   ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1268:38
     |
1268 |         try_find_fwd_imp(aut, input, NoOp, Anchored::Yes, earliest)
     |                                      ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1277:42
     |
1277 |             try_find_fwd_imp(aut, input, NoOp, Anchored::No, true)
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1279:42
     |
1279 |             try_find_fwd_imp(aut, input, NoOp, Anchored::No, false)
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1294:19
     |
1294 |     let mut mat = NoOp;
     |                   ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1303:42
     |
1303 |             Candidate::None => return Ok(NoOp),
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1428:17
     |
1428 |     state.mat = NoOp;
     |                 ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1438:50
     |
1438 |         try_find_overlapping_fwd_imp(aut, input, NoOp, state)
     |                                                  ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1485:42
     |
1485 |                 state.next_match_index = NoOp;
     |                                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1486:29
     |
1486 |                 state.mat = NoOp;
     |                             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1586:46
     |
1586 |     let mut cur: Option<(u8, u8, StateID)> = NoOp;
     |                                              ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/automaton.rs:1606:9
     |
1606 |         NoOp
     |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/dfa.rs:770:26
    |
770 |     let mut prev_class = NoOp;
    |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/contiguous.rs:823:28
    |
823 |                     return NoOp;
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/contiguous.rs:836:21
    |
836 |                     NoOp
    |                     ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/contiguous.rs:841:28
    |
841 |                     return NoOp;
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/noncontiguous.rs:581:28
    |
581 |                 prefilter: NoOp,
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/nfa/noncontiguous.rs:1243:26
     |
1243 |         QueuedSet { set: NoOp }
     |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:117:20
    |
117 |             force: NoOp,
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:118:30
    |
118 |             force_teddy_fat: NoOp,
    |                              ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:119:24
    |
119 |             force_avx: NoOp,
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:145:26
    |
145 |             self.force = NoOp;
    |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:183:26
    |
183 |             self.force = NoOp;
    |                          ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:240:20
    |
240 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/api.rs:644:20
    |
644 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/pattern.rs:213:20
    |
213 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/rabinkarp.rs:111:20
    |
111 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/rabinkarp.rs:124:24
    |
124 |                 return NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/rabinkarp.rs:164:13
    |
164 |             NoOp
    |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/teddy/mod.rs:24:13
   |
24 |             NoOp
   |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/packed/teddy/mod.rs:46:13
   |
46 |             NoOp
   |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/alphabet.rs:94:73
   |
94 |         ByteClassElementRanges { elements: self.elements(class), range: NoOp }
   |                                                                         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/alphabet.rs:154:9
    |
154 |         NoOp
    |         ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
  --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:89:32
   |
89 |             Candidate::None => NoOp,
   |                                ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:165:20
    |
165 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:283:24
    |
283 |             self.one = NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:421:13
    |
421 |             NoOp
    |             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:457:24
    |
457 |                 return NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:467:29
    |
467 |                 0 => return NoOp,
    |                             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:696:24
    |
696 |                 return NoOp;
    |                        ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:710:28
    |
710 |                     return NoOp;
    |                            ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/prefilter.rs:716:29
    |
716 |                 0 => return NoOp,
    |                             ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/primitives.rs:374:20
    |
374 |             return NoOp;
    |                    ^^^^ not found in this scope

error[E0425]: cannot find value `NoOp` in this scope
    --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/aho-corasick-1.0.1/src/util/search.rs:1105:36
     |
1105 |             MatchKind::Standard => NoOp,
     |                                    ^^^^ not found in this scope

Some errors have detailed explanations: E0408, E0416, E0425.
For more information about an error, try `rustc --explain E0408`.
error: could not compile `aho-corasick` due to 55 previous errors
error: could not compile `aho-corasick` due to 55 previous errors
error[E0425]: cannot find value `NoOp` in this scope
   --> /Users/sichanghe/.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.109/src/lib.rs:967:23
    |
967 |     let mut shebang = NoOp;
    |                       ^^^^ not found in this scope

error: could not compile `syn` due to previous error
```

</details>

To solve this, I deleted my `~/.cargo/registry`.

However, I believe that Rust-Analyzer should have prevented me from renaming `Option::None` in the first place because it is a struct defined in the standard library.

**rust-analyzer version**: rust-analyzer 0.3.1402-standalone (646f97385 2023-02-12)

**rustc version**: rustc 1.69.0 (84c898d65 2023-04-16)

### PR DESCRIPTION
**Pull request by alibektas**
fixes #14850 . This makes me wonder , why stop at structs and not do the same for other ADTs? Would be happy to add them too if nothing speaks against it.

### CODE REVIEW

**Comment by HKalbasi** [comment]
I think it should include all items (ADTs, functions, consts, ...)

File: crates/ide-db/src/rename.rs [hunk 1]
```diff
@@ -71,12 +71,25 @@ impl Definition {
         sema: &Semantics<'_, RootDatabase>,
         new_name: &str,
     ) -> Result<SourceChange> {
+        if let Some(krate) = self.krate(sema.db) {
```
Comment (lowr) [inline crates/ide-db/src/rename.rs:74]: Can you add a comment that `self.krate()` is `None` when `self` is either a built-in attribute, built-in type, or tool module, and the former two are handled below in a match?

I wonder if we should bail out here as well when the def is a tool module.

File: crates/ide-db/src/rename.rs [hunk 2]
```diff
@@ -71,12 +71,25 @@ impl Definition {
...
+            if !matches!(krate.origin(sema.db), CrateOrigin::Local { .. }) {
```
Comment (lowr) [inline crates/ide-db/src/rename.rs:75]:
```suggestion
            if !krate.origin(sema.db).is_local() {
```

File: crates/ide/src/rename.rs [hunk 3]
```diff
@@ -2487,4 +2487,18 @@ fn main() {
...
+
+    #[test]
+    fn disallow_renaming_for_non_local_definition() {
+        check(
+            "Baz",
+            r#"
+//- /lib.rs crate:lib new_source_root:library
+pub struct S$0;
+//- /main.rs crate:main deps:lib new_source_root:local
+use lib::S;
```
Comment (lowr) [inline crates/ide/src/rename.rs:2499]: I think placing `$0` at main is a more realistic test case.

Also, `new_source_root` definitely works, but I'd say the new `library` meta better fits for things we judge by `CrateOrigin` now that we have it.

File: crates/ide-db/src/rename.rs [hunk 4]
```diff
@@ -71,12 +71,25 @@ impl Definition {
...
         match *self {
             Definition::Module(module) => rename_mod(sema, module, new_name),
             Definition::BuiltinType(_) => {
                 bail!("Cannot rename builtin type")
             }
             Definition::SelfType(_) => bail!("Cannot rename `Self`"),
+            Definition::Macro(mac) => {
+                if mac.is_builtin_derive(sema.db) {
+                    bail!("Cannot rename builtin macro")
+                }
```
Comment (lowr) [inline crates/ide-db/src/rename.rs:89]: Can you also add a test for this?

Comment (alibektas) [reply crates/ide-db/src/rename.rs:89]: I can't seem to find a way that would bypass `Cannot rename a non-local definition.` and bail out at `Cannot rename builtin derive`

File: crates/ide-db/src/rename.rs [hunk 5]
```diff
@@ -71,12 +71,35 @@ impl Definition {
...
+                bail!("Cannot rename a builtin attr.")
+            }
             Definition::SelfType(_) => bail!("Cannot rename `Self`"),
+            Definition::Macro(mac) => {
+                if mac.is_builtin_derive(sema.db) {
+                    bail!("Cannot rename builtin derive")
+                }
+
+                rename_reference(sema, Definition::Macro(mac), new_name)
+            }
```
Comment (Veykril) [inline crates/ide-db/src/rename.rs:102]: This isn't necessary, all builtin derives have a macro assocaited with them that lives in core/std so this path won't ever be hit iven the earlier check

**Review by Veykril — APPROVED** [review]

**Pull request merged**